nvflare/app_common/aggregators/async.py

```diff
@@ -0,0 +1,176 @@
+            self.log_debug(fl_ctx, "done adding contribution to queue")
+            return True
+        else:
+            # when the aggregation is started, we no longer accept new contributions
+            self.log_warning(fl_ctx, "dropped contribution while aggregating")
+            return False
+
+    def _do_accept(self):
+        self.logger.debug("Started accept thread")
+        while True:
+            if self.run_ended:
+                # the job is already done or aborted
+                self.logger.debug("run ended - exit")
+                break
+
+            try:
+                # we wait very shortly when trying to get a contribution, so we can check other conditions
+                contrib = self.contributions.get(timeout=_SHORT_WAIT)
+            except queue.Empty:
+                contrib = None
+
+            if contrib:
+                assert isinstance(contrib, _Contribution)
+                self.log_debug(contrib.fl_ctx, "Accepting contribution")
+                try:
+                    accepted = self.aggregator.accept(contrib.data, contrib.fl_ctx)
+                    self.log_debug(contrib.fl_ctx, f"{type(self.aggregator)} processed contribution: {accepted=}")
+                except Exception as ex:
+                    self.log_exception(
+                        contrib.fl_ctx,
+                        f"exception from {type(self.aggregator)} when accept: {secure_format_exception(ex)}",
+                    )
+
+            if self.aggregating and self.contributions.empty() and not self.accept_done.is_set():
+                # When self.aggregating is set, the "aggregate" process is started and waiting for all contributions
+                # to be accepted.
+                # We set "accept_done" when all pending contributions are done.
+                self.logger.debug("Finished accept for one round")
+                self.accept_done.set()
+
+        self.logger.debug("Finished accept thread")
+
+    def aggregate(self, fl_ctx: FLContext) -> Shareable:
+        self.log_debug(fl_ctx, "starting to aggregate")
+        # set self.aggregating to notify the "accept" thread that we are starting aggregation.
+        self.aggregating = True
+
+        # we wait until the "accept" thread is done with pending contributions
+        rc = conditional_wait(
+            waiter=self.accept_done,
+            timeout=self.accept_timeout,
+            abort_signal=fl_ctx.get_run_abort_signal(),
+            condition_cb=self._check_end_run,
+        )
+
+        if rc in [_AcceptWaitRC.ABORTED, _AcceptWaitRC.END_RUN]:
+            self.log_info(fl_ctx, "skipped aggregation since job is aborted")
+            return Shareable()
+
+        if rc != _AcceptWaitRC.IS_SET:
+            self.log_warning(fl_ctx, f"abnormal result {rc} waiting for accept thread")
+
+        # we then call the aggregator to perform actual aggregation
+        result = self.aggregator.aggregate(fl_ctx)
```

Comment:
do we need try-except block here?

Reply:
I intentionally did not use try-except block to let the exception pop up and to let upper layer handle it.
In case of SAG, this would cause the workflow to end, which is fine.